com.onesignal.unity.android/Runtime/AndroidNotificationsManager.cs

```diff
@@ -0,0 +1,151 @@
+            }
+
+            /// <param name="permission">boolean</param>
+            public void onPermissionChanged(bool permission) {
+                UnityMainThreadDispatch.Post(state => _parent.PermissionChanged?.Invoke(permission));
+            }
+        }
+
+        private sealed class INotificationWillShowInForegroundHandler : OneSignalAndroidJavaProxy {
+            private AndroidNotificationsManager _parent;
+
+            public INotificationWillShowInForegroundHandler(AndroidNotificationsManager notificationsManager) : base("notifications.INotificationWillShowInForegroundHandler") {
+                _parent = notificationsManager;
+            }
+
+            /// <param name="notificationReceivedEvent">INotificationReceivedEvent</param>
+            public void notificationWillShowInForeground(AndroidJavaObject notificationReceivedEvent) {
+                var notifJO = notificationReceivedEvent.Call<AndroidJavaObject>("getNotification");
+
+                if (_parent.WillShow == null) {
+                    notificationReceivedEvent.Call("complete", notifJO);
+                    return;
+                }
+
+                var notification = _getNotification(notifJO);
+
+                UnityMainThreadDispatch.Post(state => notificationReceivedEvent.Call("complete", _parent.WillShow(notification) != null ? notifJO : null));
+            }
+        }
+
+        private sealed class INotificationClickHandler : OneSignalAndroidJavaProxy {
+            private AndroidNotificationsManager _parent;
+
+            public INotificationClickHandler(AndroidNotificationsManager notificationsManager) : base("notifications.INotificationClickHandler") {
+                _parent = notificationsManager;
+            }
+
+            /// <param name="result">INotificationClickResult</param>
+            public void notificationClicked(AndroidJavaObject result) {
+                var notifJO = result.Call<AndroidJavaObject>("getNotification");
+                var notification = _getNotification(notifJO);
+
+                var actionJO = result.Call<AndroidJavaObject>("getAction");
+                var action = _getAction(actionJO);
+
+                var notifClickResult = new NotificationClickedResult {
+                    notification = notification,
+                    action = action
+                };
+
+                UnityMainThreadDispatch.Post(state => _parent.Clicked?.Invoke(notifClickResult));
+            }
+        }
+
+        private static Notification _getNotification(AndroidJavaObject notifJO) {
+            var notification = notifJO.ToSerializable<Notification>();
+            
+            var dataJson = notifJO.Call<AndroidJavaObject>("getAdditionalData");
+            if (dataJson != null) {
+                var dataJsonStr = dataJson.Call<string>("toString");
+                notification.additionalData = Json.Deserialize(dataJsonStr) as Dictionary<string, object>;
```

Comment:
FYI I'm considering changing `additionalData` on Android to be of type Dictionary<string, Any>. If this does change, this will also need to change.